internal/controller/object/informers.go

```diff
@@ -0,0 +1,283 @@
+}
+
+type resourceCache struct {
+	cache    cache.Cache
+	cancelFn context.CancelFunc
+}
+
+var _ source.Source = &resourceInformers{}
+
+// Start implements source.Source, i.e. starting resourceInformers as
+// source with h as the sink of update events. It keeps sending events until
+// ctx is done.
+func (i *resourceInformers) Start(ctx context.Context, h handler.EventHandler, q workqueue.RateLimitingInterface, ps ...predicate.Predicate) error {
+	if i.sink != nil {
+		return errors.New("source already started, cannot start it again")
+	}
+	i.sink = func(providerConfig string, ev runtimeevent.GenericEvent) {
+		for _, p := range ps {
+			if !p.Generic(ev) {
+				return
+			}
+		}
+		h.Generic(context.WithValue(ctx, keyProviderConfigName, providerConfig), ev, q)
+	}
+
+	go func() {
+		<-ctx.Done()
+		i.sink = nil
+	}()
+
+	return nil
+}
+
+// WatchResources starts informers for the given resource GVKs for the given
+// cluster (i.e. rest.Config & providerConfig).
+// The is wired into the Object reconciler, which will call this method on
+// every reconcile to make resourceInformers aware of the referenced or managed
+// resources of the given Object.
+//
+// Note that this complements garbageCollectResourceInformers which regularly
+// garbage collects resource informers that are no longer referenced by
+// any Object.
+func (i *resourceInformers) WatchResources(rc *rest.Config, providerConfig string, gvks ...schema.GroupVersionKind) {
+	if rc == nil {
+		rc = i.config
+	}
+
+	// start new informers
+	for _, gvk := range gvks {
+		i.lock.RLock()
+		_, found := i.resourceCaches[gvkWithConfig{providerConfig: providerConfig, gvk: gvk}]
+		i.lock.RUnlock()
+		if found {
+			continue
+		}
+
+		log := i.log.WithValues("providerConfig", providerConfig, "gvk", gvk.String())
+
+		ca, err := cache.New(rc, cache.Options{})
+		if err != nil {
+			log.Debug("failed creating a cache", "error", err)
+			continue
+		}
+
+		// don't forget to call cancelFn in error cases to avoid leaks. In the
+		// happy case it's called from the go routine starting the cache below.
+		ctx, cancelFn := context.WithCancel(context.Background())
+
+		u := kunstructured.Unstructured{}
+		u.SetGroupVersionKind(gvk)
+		inf, err := ca.GetInformer(ctx, &u, cache.BlockUntilSynced(false)) // don't block. We wait in the go routine below.
+		if err != nil {
+			cancelFn()
+			log.Debug("failed getting informer", "error", err)
+			continue
+		}
+
+		if _, err := inf.AddEventHandler(kcache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				ev := runtimeevent.GenericEvent{
+					Object: obj.(client.Object),
+				}
+
+				i.sink(providerConfig, ev)
+			},
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				old := oldObj.(client.Object) //nolint:forcetypeassert // Will always be client.Object.
+				obj := newObj.(client.Object) //nolint:forcetypeassert // Will always be client.Object.
+				if old.GetResourceVersion() == obj.GetResourceVersion() {
```

Comment:
why is this good?

Reply:
I believe it is good since there is no change in the resource, so no need to trigger a new reconcile. Not sure how/why this could happen though - e.g. resource didn't change but an update event thrown.

Same as https://github.com/crossplane/crossplane/blob/f626e8a530e2ecaecb7cd59633f0871836e4aa1c/internal/controller/apiextensions/definition/composed.go#L170.

Reply:
Ok, removed that check.